receivers/discord/discord.go

```diff
@@ -202,52 +201,62 @@ func (d Notifier) SendResolved() bool {
 	return !d.GetDisableResolveMessage()
 }
 
-func (d Notifier) constructAttachments(ctx context.Context, as []*types.Alert, embedQuota int) []discordAttachment {
-	attachments := make([]discordAttachment, 0)
+func (d Notifier) constructAttachments(ctx context.Context, alerts []*types.Alert, embedQuota int) []discordAttachment {
+	attachments := make([]discordAttachment, 0, embedQuota)
+	for _, alert := range alerts {
+		// Check if the image limit has been reached at the start of each iteration.
+		if embedQuota < 1 {
+			break
+		}
 
-	_ = images.WithStoredImages(ctx, d.log, d.images,
-		func(index int, image images.Image) error {
-			if embedQuota < 1 {
-				return images.ErrImagesDone
+		attachment, err := d.getAttachmentFromURL(ctx, alert)
+		if err != nil {
+			if !errors.Is(err, images.ErrImagesNoURL) {
+				// If there's an error, continue iterating through the other alerts.
+				continue
 			}
-
-			if len(image.URL) > 0 {
-				attachments = append(attachments, discordAttachment{
-					url:       image.URL,
-					state:     as[index].Status(),
-					alertName: as[index].Name(),
-				})
-				embedQuota--
-				return nil
+			// The image has no public URL, use the bytes for the attachment.
+			attachment, err = d.getAttachmentFromBytes(ctx, alert)
+			if err != nil {
+				continue
 			}
+		}
 
-			// If we have a local file, but no public URL, upload the image as an attachment.
-			if len(image.Path) > 0 {
-				base := filepath.Base(image.Path)
-				url := fmt.Sprintf("attachment://%s", base)
-				reader, err := images.OpenImage(image.Path)
-				if err != nil && !errors.Is(err, images.ErrImageNotFound) {
-					d.log.Warn("failed to retrieve image data from store", "error", err)
-					return nil
-				}
-
-				attachments = append(attachments, discordAttachment{
-					url:       url,
-					name:      base,
-					reader:    reader,
-					state:     as[index].Status(),
-					alertName: as[index].Name(),
-				})
-				embedQuota--
-			}
-			return nil
-		},
-		as...,
-	)
+		// We got an attachment, either using the image URL or bytes.
+		attachments = append(attachments, attachment)
+		embedQuota--
+	}
 
 	return attachments
 }
 
+func (d Notifier) getAttachmentFromURL(ctx context.Context, alert *types.Alert) (discordAttachment, error) {
+	url, err := d.images.GetImageURL(ctx, alert)
```

Comment:
How did we locally test this change without Grafana having been updated to use the latest `alerting` version? The live image store in grafana does not have any working implementation of `GetImageURL`. Is there a separate PR or branch for that?

Reply:
> Is there a separate PR or branch for that?

Not yet, the plan was to implement that in `grafana/grafana` next.